typedData/typedData.go

```diff
@@ -0,0 +1,905 @@
+
+type TypeDefinition struct {
+	Name               string     `json:"-"`
+	Enconding          *felt.Felt `json:"-"`
+	EncoddingString    string     `json:"-"`
+	SingleEncString    string     `json:"-"`
+	ReferencedTypesEnc []string   `json:"-"`
+	Parameters         []TypeParameter
+}
+
+type TypeParameter struct {
+	Name     string `json:"name"`
+	Type     string `json:"type"`
+	Contains string `json:"contains,omitempty"`
+}
+
+// NewTypedData creates a new instance of TypedData.
+//
+// Parameters:
+// - types: a slice of TypeDefinition representing the types used in the TypedData.
+// - primaryType: a string representing the primary type of the TypedData.
+// - domain: a Domain struct representing the domain information of the TypedData.
+// - message: a byte slice containing the JSON-encoded message.
+//
+// Returns:
+// - td: a pointer to the newly created TypedData instance.
+// - err: an error if any occurred during the creation of the TypedData.
+func NewTypedData(types []TypeDefinition, primaryType string, domain Domain, message []byte) (td *TypedData, err error) {
+	//types
+	typesMap := make(map[string]TypeDefinition)
+	for _, typeDef := range types {
+		typesMap[typeDef.Name] = typeDef
+	}
+
+	//primary type
+	if _, ok := typesMap[primaryType]; !ok {
+		return td, fmt.Errorf("invalid primary type: %s", primaryType)
+	}
+
+	//message
+	messageMap := make(map[string]any)
+	err = json.Unmarshal(message, &messageMap)
+	if err != nil {
+		return td, fmt.Errorf("error unmarshalling the message: %w", err)
+	}
+
+	//revision
+	revision, err := GetRevision(domain.Revision)
+	if err != nil {
+		return td, fmt.Errorf("error getting revision: %w", err)
+	}
+
+	//domain type encoding
+	domainTypeDef, err := encodeTypes(revision.Domain(), typesMap, revision)
+	if err != nil {
+		return td, err
+	}
+	typesMap[revision.Domain()] = domainTypeDef
+
+	//types encoding
+	primaryTypeDef, err := encodeTypes(primaryType, typesMap, revision)
+	if err != nil {
+		return td, err
+	}
+	typesMap[primaryType] = primaryTypeDef
+
+	for _, typeDef := range typesMap {
+		if typeDef.EncoddingString == "" {
+			return td, fmt.Errorf("'encodeTypes' failed: type '%s' doesn't have encode value", typeDef.Name)
+		}
+	}
+
+	td = &TypedData{
+		types:       typesMap,
+		primaryType: primaryType,
+		domain:      domain,
+		message:     messageMap,
+		revision:    revision,
+	}
+
+	return td, nil
+}
+
+// GetMessageHash calculates the hash of a typed message for a given account using the StarkCurve.
+//
+// (ref: https://github.com/starknet-io/SNIPs/blob/5d5a42c654c27b377d8b7f90b453065fd19ec2eb/SNIPS/snip-12.md#specification)
+//
+// Parameters:
+// - account: A string representing the account.
+// Returns:
+// - hash: A pointer to a felt.Felt representing the calculated hash.
+func (td *TypedData) GetMessageHash(account string) (hash *felt.Felt, err error) {
+	//signed_data = encode(PREFIX_MESSAGE, Enc[domain_separator], account, Enc[message])
+
+	elements := []*felt.Felt{}
```

Comment:
nit: we know the size of elements beforehand, so cna pre-define its size. We could also just get rid of it and pass in the paramters directly.

Reply:
makes total sense haha. I'll do it